src/javascript/crypto/e2e/coname/lookup_verify.js

```diff
@@ -0,0 +1,602 @@
+ */
+e2e.coname.verifyQuorumSignature_ = function(pk, message, sig) {
+  var pubKeyType = Object.keys(pk)[0];
+
+  switch (pubKeyType) {
+    case 'ed25519':
+      return new e2e.ecc.Ed25519(e2e.ecc.PrimeCurve.ED_25519, {
+        'pubKey': pk[pubKeyType]
+      }).verify(message, sig);
+  }
+
+  return false;
+};
+
+
+/**
+ * @private
+ * @param {object} rcg The RealmConfig
+ * @param {object} ratifications The []*proto.SignedEpochHead
+ * @param {Date} now The current time
+ * @return {!e2e.ByteArray} the root hash
+ */
+e2e.coname.verifyConsensus_ = function(rcg, ratifications, now) {
+  var i = 1, n = ratifications.length, want, got, pks, have = {}, t, valid,
+      verifyQuorumSignature_ = e2e.coname.verifyQuorumSignature_;
+
+  if (n === 0) {
+    throw new e2e.error.InvalidArgumentsError(
+        'VerifyConsensus: no signed epoch heads provided');
+  }
+  // check that all the SEHs have the same head
+  for (; i < n; i++) {
+    want = ratifications[0].head.head.encoding;
+    got = ratifications[i].head.head.encoding;
+    if (!e2e.compareByteArray(want, got)) {
+      throw new e2e.error.InvalidArgumentsError(
+          'VerifyConsensus: epoch heads don\'t match: ' + want + ' vs ' + got);
+    }
+  }
+  // check that the seh is not expired
+  t = 1000 * (ratifications[0].head.head.issue_time.seconds +
+      rcg.epoch_time_to_live.seconds);
+  if (now.getTime() > t) {
+    n = new Date();
+    n.setTime(t);
+    throw new e2e.error.InvalidArgumentsError(
+        'VerifyConsensus: epoch expired at ' + n + ' < ' + now);
+  }
+
+  // check that there are sufficiently many fresh signatures.
+  pks = rcg.verification_policy.public_keys;
+  want = rcg.verification_policy.quorum;
+  valid = goog.array.some(e2e.coname.flattenQuorum_(want), function(id) {
+    goog.array.some(ratifications, function(seh) {
+      var sig = seh.signatures[id];
+      return (sig && verifyQuorumSignature_(pks[id], seh.head.encoding, sig)) ?
+          (have[id] = true) :
+          false;
+    });
+    return e2e.coname.checkQuorum_(want, have);
+  });
+
+  if (valid) {
+    return ratifications[0].head.head.root_hash;
+  }
+
+  throw new e2e.error.InvalidArgumentsError(
+      'VerifyConsensus: insufficient signatures (have ' +
+      JSON.stringify(have) + ', want ' +
+      JSON.stringify(want) + ')');
+};
+
+
+/**
+ * Converts a non-negative 32-bit integer into a 4-byte little-endian ByteArray
+ * @private
+ * @param {!number} value The number to convert
+ * @return {!e2e.ByteArray} The number as a little-endian ByteArray.
+ */
+e2e.coname.numberTo4ByteArray_ = function(value) {
+  if (value < 0 || value > 4294967295) {
+    throw new e2e.error.InvalidArgumentsError('The number is out of range');
+  }
+
+  for (var i = 0, byteArray = [0, 0, 0, 0]; value > 0; i++) {
+    byteArray[i] = value & 0xff;
+    value >>>= 8;
+  }
+  return byteArray;
+};
+
+
+/**
+ * @private
+ * assumes ownership of the array underlying prefixBits
```

Comment:
I think this comment is incorrect here since arrays work differently in JavaScript. In Go, using `append` mutates the  array underlying a slice, but in JavaScript, using `concat` creates a new array. Thus, I think `prefixBits` isn't modified at all in this version of the code, so it doesn't actually need to "assume ownership".